kotest-tests/kotest-tests-core/src/jvmTest/kotlin/com/sksamuel/kotest/core/runtime/parallel/SingleInstanceTest.kt

```diff
@@ -0,0 +1,291 @@
+package com.sksamuel.kotest.core.runtime.parallel
+
+import com.sksamuel.kotest.core.runtime.PersistentThreadLocal
+import io.kotest.assertions.assertSoftly
+import io.kotest.assertions.throwables.shouldThrow
+import io.kotest.core.spec.IsolationMode
+import io.kotest.core.spec.style.FunSpec
+import io.kotest.matchers.ints.shouldBeExactly
+import io.kotest.matchers.ints.shouldBeGreaterThan
+import io.kotest.matchers.maps.shouldHaveSize
+import io.kotest.matchers.shouldBe
+import kotlinx.coroutines.delay
+import java.util.concurrent.atomic.AtomicInteger
+import java.util.concurrent.locks.ReentrantLock
+import kotlin.concurrent.getOrSet
+
+class SpecThreadSingleInstanceTest : FunSpec({
+
+   isolation = IsolationMode.SingleInstance
+   threadsForSpec = 3
+
+   val multipleThreadMultipleInvocationCounter =
+      PersistentThreadLocal<Int>()
+
+   afterSpec {
+      multipleThreadMultipleInvocationCounter.map.shouldHaveSize(3)
+      multipleThreadMultipleInvocationCounter.map.values.sum() shouldBe 3
+   }
+
+   test("test 1 should create own key in map with value 1") {
+      val counter = multipleThreadMultipleInvocationCounter.getOrSet { 0 }
+      multipleThreadMultipleInvocationCounter.set(counter + 1)
+   }
+
+   test("test 2 should create own key in map with value 1") {
+      val counter = multipleThreadMultipleInvocationCounter.getOrSet { 0 }
+      multipleThreadMultipleInvocationCounter.set(counter + 1)
+   }
+
+   test("test 3 should create own key in map with value 1") {
+      val counter = multipleThreadMultipleInvocationCounter.getOrSet { 0 }
+      multipleThreadMultipleInvocationCounter.set(counter + 1)
+   }
+
+   val lock = ReentrantLock()
+
+   test("test should lock object") {
+      lock.lock()
+      delay(1000)
+      lock.unlock()
+   }
+
+   test("lock should be locked") {
+      delay(300)
+      lock.isLocked shouldBe true
+   }
+
+   test("lock should be locked too") {
+      delay(300)
+      shouldThrow<AssertionError> {
+         lock.isLocked shouldBe false
+      }
+   }
+})
+
+class SpecThreadWithNestedTestSingleInstanceTest : FunSpec({
+
+   isolation = IsolationMode.SingleInstance
+   threadsForSpec = 3
+
+   val multipleThreadCounter =
+      PersistentThreadLocal<Int>()
+
+   afterSpec {
+      assertSoftly {
+         multipleThreadCounter.map.shouldHaveSize(2)
+         multipleThreadCounter.map.values.sum() shouldBe 4
```

Comment:
This doesn't assert which threads added what.
I would append characters to a string, and then assert that unique string forms are made.

Reply:
I guess it would be possible for the first test to run twice, and the second test not to run at all, and you'd still get the answer = 2.
If you do + "a" and + "b" then you know you will have "ab". It's safer.

Reply:
Done